benchmarks/distributed/ddp/compare/python_ddp.py

```diff
@@ -0,0 +1,228 @@
+        assert buffer_size > max_numel, "buffer_size: {} should be larger than largest param: {}".format(buffer_size, max_numel)
+
+        # Build buckets for params
+        self.param_to_bucket, self.buckets = self._build_buckets_for_params(buffer_size)
+
+        # Register per-parameter hook to be invoked when grad is ready.
+        for p in self.module.parameters():
+            assert p.requires_grad
+            p.register_hook(functools.partial(self._on_param_grad_ready, p))
+
+    def _build_buckets_for_params(self, max_buffer_size):
+        """
+        Distributes params into list of buckets. Maintains param -> bucket
+        mapping. Returns tuple of (param_to_buckets, buckets).
+        """
+        print("_build_buckets_for_params called")
+        params_to_buckets = {}
+        buckets = set()
+        cur_bucket = self.Bucket(max_buffer_size)
+        total_param = 0
+        for param in self.module.parameters():
+            total_param += 1
+            assert param.requires_grad, "param.requires_grad must be True"
+            if cur_bucket.try_hold_param(param):
+                params_to_buckets[param] = cur_bucket
+                buckets.add(cur_bucket)
+            else:
+                new_bucket = self.Bucket(max_buffer_size)
+                assert new_bucket.try_hold_param(param), "param must be holded in a empty bucket"
+                params_to_buckets[param] = new_bucket
+                buckets.add(new_bucket)
+                cur_bucket = new_bucket
+
+        first_param = next(self.module.parameters())
+        for bucket in buckets:
+            bucket.buffer = first_param.new(bucket.total_elements)
+            assert bucket.buffer is not None, 'bucket.buffer should not be None'
+        print("len(param_to_bucket)={}, len(buckets)={}".format(
+            len(params_to_buckets), len(buckets)))
+
+        # Sanity check to ensure all params are distributed correctly into buckets
+        total_params_in_buckets = 0
+        for bucket in buckets:
+            total_params_in_buckets += len(bucket.param_to_offset)
+        assert total_param == total_params_in_buckets
+
+        return params_to_buckets, buckets
+
+    # Callback when param.grad is ready. Note during callback, param.grad won't
+    # be ready yet, we MUST use the given ''grad'' which would be passed upon
+    # callback.
+    def _on_param_grad_ready(self, param, grad):
+        """
+        Callback when grad for param is ready. Copy grad to its corresponding
+        bucket. When the bucket is full, kickoff an async all_reduce if
+        async_reduction is set, and adds the resultant handle to
+        self.async_handles.
+
+        .. warning::
+            Note param.grad isn't set yet. Use the passed grad instead.
+        """
+        # Validate bucket and offset are set.
+        bucket = self.param_to_bucket.get(param)
+        assert bucket is not None, "Failed to find bucket for param"
```

Comment:
The `assert` in this method is also called in a loop.

Reply:
> @bowangbj Did you get a chance to collect the traces to compare the trainings using DDP core and Python Async DDP?

Thanks Yi for all the great comments. Sorry for the long delay. Finally I landed in PyTorch team after bootcamp :) 
Will follow up and dive into more details as part of my onboarding plan in Aug (thanks to Shen, Yi, Boris and Nikhil for the great proposal). 

Stay tuned!